ProcessGroupNCCL should respect timeout passed in to init_process_group.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27224 ProcessGroupNCCL should respect timeout passed in to init_process_group.**

As part of adding error handling to NCCL, we are now able to specify a
timeout for operations using ProcessGroupNCCL. Although, this timeout had a
default of 10 seconds and didn't respect the timeout specified in
init_process_group.

In this change, I've ensured we pass the appropriate timeout to
ProcessGroupNCCL.

Differential Revision: [D17717992](https://our.internmc.facebook.com/intern/diff/D17717992/)

Commit:
ProcessGroupNCCL should respect timeout passed in to init_process_group.

As part of adding error handling to NCCL, we are now able to specify a
timeout for operations using ProcessGroupNCCL. Although, this timeout had a
default of 10 seconds and didn't respect the timeout specified in
init_process_group.

In this change, I've ensured we pass the appropriate timeout to
ProcessGroupNCCL.

Differential Revision: [D17717992](https://our.internmc.facebook.com/intern/diff/D17717992/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -346,7 +346,9 @@ def init_process_group(backend,
-            This is only applicable for the ``gloo`` backend.
+            This is applicable for the ``gloo`` backend. For ``nccl``, this is
+            applicable only if the environment variable NCCL_BLOCKING_WAIT is
+            set to 1
```
nit: missing period.

mrshenli commented on test/test_c10d.py:
```diff
@@ -3098,6 +3098,31 @@ def test_invalid_nccl_blocking_wait_env(self):
+        c10d.distributed_c10d.init_process_group(
+            backend=dist.Backend.NCCL, store=store, world_size=2, rank=self.rank,
+            timeout=timedelta(seconds=timeout))
+        pg = c10d.distributed_c10d._get_default_group()
```
question: why getting the default group instead of using `c10d.all_reduce`?

pritamdamania87 replied:
Good point, let me change this.

mrshenli commented on test/test_c10d.py:
```diff
@@ -3098,6 +3098,31 @@ def test_invalid_nccl_blocking_wait_env(self):
 
+    @requires_nccl()
+    @skip_if_not_multigpu
+    def test_nccl_timeout(self):
```
Any reason for adding this to `CommTest` instead of `TimeoutTest`?

pritamdamania87 replied:
`TimeoutTest` is not a `MultiProcessTestCase`.

pietern replied:
I know the other NCCL tests are here as well, but reiterating that they all should be moved to a separate NCCL class instead. The CommTest class was intended to only hold tests for functions in `csrc/distributed/c10d/comm.cpp`.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -346,7 +346,9 @@ def init_process_group(backend,
             the process group. Default value equals 30 minutes.
-            This is only applicable for the ``gloo`` backend.
+            This is applicable for the ``gloo`` backend. For ``nccl``, this is
+            applicable only if the environment variable NCCL_BLOCKING_WAIT is
```
Q: Why does NCCL_BLOCKING_WAIT have to be set to 1 for the timeout to be applicable?

pritamdamania87 replied:
Without `NCCL_BLOCKING_WAIT` work->wait() does not block and there isn't really something that the user is waiting on.

pietern replied:
` ``NCCL_BLOCKING_WAIT`` `

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -346,7 +346,9 @@ def init_process_group(backend,
         timeout (timedelta, optional): Timeout for operations executed against
             the process group. Default value equals 30 minutes.
-            This is only applicable for the ``gloo`` backend.
+            This is applicable for the ``gloo`` backend. For ``nccl``, this is
```
nit, I think the documentation on line 132 of this can also be updated now?

Commit:
Update on "ProcessGroupNCCL should respect timeout passed in to init_process_group."


As part of adding error handling to NCCL, we are now able to specify a
timeout for operations using ProcessGroupNCCL. Although, this timeout had a
default of 10 seconds and didn't respect the timeout specified in
init_process_group.

In this change, I've ensured we pass the appropriate timeout to
ProcessGroupNCCL.

Differential Revision: [D17717992](https://our.internmc.facebook.com/intern/diff/D17717992/)

[ghstack-poisoned]

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -129,7 +129,8 @@ class GroupMember(object):
 
 # Default process group wide timeout, if applicable.
-# This currently only applies to the gloo backend. To make an attempt at
+# This currently only to the gloo backend and nccl backend
```
"This only applies to the gloo and nccl backends."

pietern commented on test/test_c10d.py:
```diff
@@ -3098,6 +3098,29 @@ def test_invalid_nccl_blocking_wait_env(self):
+
+            total_time = time.time() - start
+
+            self.assertLess(abs(total_time - timeout), 0.5)
```
Why assert less than 0.5 if the timeout is 1 second?

To really test the timeout path (instead of the other process terminating), rank 1 should sleep until rank 0 has timed out.

pritamdamania87 replied:
0.5 is just the leeway not the actual timeout. Ideally total_time - timeout should be 0, but there would be some overhead I'm accounting for and being conservative with 0.5 to avoid this test being flaky.

Commit:
Update on "ProcessGroupNCCL should respect timeout passed in to init_process_group."


As part of adding error handling to NCCL, we are now able to specify a
timeout for operations using ProcessGroupNCCL. Although, this timeout had a
default of 10 seconds and didn't respect the timeout specified in
init_process_group.

In this change, I've ensured we pass the appropriate timeout to
ProcessGroupNCCL.

Differential Revision: [D17717992](https://our.internmc.facebook.com/intern/diff/D17717992/)

[ghstack-poisoned]

mrshenli commented on test/test_c10d.py:
```diff
@@ -3098,6 +3069,81 @@ def test_invalid_nccl_blocking_wait_env(self):
+            self.assertLess(abs(total_time - timeout), 0.5)
+        else:
+            # Ensure the other rank sleeps to trigger timeout.
+            time.sleep(2 * timeout)
```
What error do we see if peers don't sleep here? An immediate "connection closed by peer" error?

pritamdamania87 replied:
I added this part as per Pieter's suggestion. If we don't sleep, we still do see a timeout in this test case. Basically, the test works if the other rank sleeps/exits. I think Pieter's suggestion above to use sleep() was to make it more realistic in terms of how a timeout usually happens.
